### PR TITLE
geo: fix st_segmentize to work with ZM coords

### LINKED ISSUE #62290: spatial: st_segmentize goes out of bounds on Z/M coords
[(roachtest).sqlsmith/setup=tpcc/setting=default failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2795641&tab=buildLog) on [master@3d19b2cf6b290a152b23722fc32e995eed3b437b](https://github.com/cockroachdb/cockroach/commits/3d19b2cf6b290a152b23722fc32e995eed3b437b):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=tpcc/setting=default/run_1
	sqlsmith.go:218,sqlsmith.go:253,test_runner.go:768: error: pq: internal error: runtime error: index out of range [15] with length 15
		stmt:
		WITH
			with_29 (col_243)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(e'\x04*\x04':::STRING),
								(e'Zk\x0f"\r':::STRING),
								((NULL::TIMESTAMP || '':::STRING::STRING)::STRING),
								('|e':::STRING)
						)
							AS tab_125 (col_243)
				),
			with_30 (col_244) AS (SELECT * FROM (VALUES (2406307256:::OID), (NULL)) AS tab_126 (col_244))
		SELECT
			count(*) AS col_245
		FROM
			defaultdb.public.order_line@order_line_stock_fk_idx AS tab_127
		WHERE
			NULL
		GROUP BY
			tab_127.ol_supply_w_id
		HAVING
			st_intersects('010100008008F544DEA68002C24848B7372B3EFAC1B4D701DBF187EC41':::GEOMETRY::GEOMETRY, st_segmentize('010500004002000000010200004003000000A8C29CD554F3D2418853FF97E5C5FAC10B8156ABEDA7F3C1E8683C1DE252EF413739717AD77100C21C2AB284E605F44100E8746E29548C410099F27AF096E4418C98902C8018F8C1010200004002000000B267F30BEC7CF3C1601A79C86681E6415CF936003112FB41F3532DB9B9DE01C2F092845CF7BDD541BA662D74E55BF541':::GEOMETRY::GEOMETRY, 1.401298464324817e-45:::FLOAT8::FLOAT8)::GEOMETRY::GEOMETRY)::BOOL
		LIMIT
			58:::INT8;

```

<details><summary>More</summary><p>

Artifacts: [/sqlsmith/setup=tpcc/setting=default](https://teamcity.cockroachdb.com/viewLog.html?buildId=2795641&tab=artifacts#/sqlsmith/setup=tpcc/setting=default)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dtpcc%2Fsetting%3Ddefault.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

### PR DESCRIPTION
**Pull request by andyyang890**
Previously, st_segmentize did not handle the Z and M
coordinate properly.

Fixes #62290.

Release note: None

### CODE REVIEW

File: pkg/geo/geomfn/segmentize.go [hunk 1]
```diff
@@ -67,14 +75,19 @@ func segmentizeCoords(a geom.Coord, b geom.Coord, maxSegmentLength float64) ([]f
...
-	allSegmentizedCoordinates := make([]float64, 0, 2*(1+numberOfSegmentsToCreate))
+	allSegmentizedCoordinates := make([]float64, 0, numPoints)
 	allSegmentizedCoordinates = append(allSegmentizedCoordinates, a.Clone()...)
 	segmentFraction := 1.0 / float64(numberOfSegmentsToCreate)
 	for pointInserted := 1; pointInserted < numberOfSegmentsToCreate; pointInserted++ {
+		segmentPoint := make([]float64, 0, len(a))
+		for i := 0; i < len(a); i++ {
+			segmentPoint = append(
+				segmentPoint,
+				a[i]*(1-float64(pointInserted)*segmentFraction)+b[i]*(float64(pointInserted)*segmentFraction))
```
Comment (otan) [inline pkg/geo/geomfn/segmentize.go:86]: nit: trailing comma
```
				a[i]*(1-float64(pointInserted)*segmentFraction)+b[i]*(float64(pointInserted)*segmentFraction),
)
```

File: pkg/geo/geomfn/segmentize.go [hunk 2]
```diff
@@ -51,12 +51,20 @@ func Segmentize(g geo.Geometry, segmentMaxLength float64) (geo.Geometry, error)
 // segment has a length less than or equal to given maximum segment length.
 // Note: List of points does not consist of end point.
 func segmentizeCoords(a geom.Coord, b geom.Coord, maxSegmentLength float64) ([]float64, error) {
+	if len(a) != len(b) {
```
Comment (otan) [inline pkg/geo/geomfn/segmentize.go:54]: can you test these error cases in the unit tests? look at other places we do `errorTestCases`

**Review by otan — APPROVED** [review]

**Comment by andyyang890** [comment]
I realized I didn't fix the `segmentizeCoords` function for the geography variant so will update the PR once I fix that

File: pkg/geo/geogfn/segmentize.go [hunk 3]
```diff
@@ -65,19 +72,17 @@ func segmentizeCoords(a geom.Coord, b geom.Coord, segmentMaxAngle float64) ([]fl
...
-	// 1 / (2^n)[numberOfSegmentsToCreate] <= segmentMaxLength / distanceBetweenPoints < 1 / (2^(n-1))
-	// (2^n)[numberOfSegmentsToCreate] >= distanceBetweenPoints / segmentMaxLength > 2^(n-1)
-	// therefore n = ceil(log2(segmentMaxLength/distanceBetweenPoints)). Hence
-	// numberOfSegmentsToCreate = 2^(ceil(log2(segmentMaxLength/distanceBetweenPoints))).
+	// We can write that power as 2^n in the following inequality
+	// 2^n >= ceil(chordAngleBetweenPoints/segmentMaxLength) > 2^(n-1).
+	// We can drop the ceil since 2^n must be an int
+	// 2^n >= chordAngleBetweenPoints/segmentMaxLength > 2^(n-1).
+	// Then n = ceil(log2(chordAngleBetweenPoints/segmentMaxLength)).
+	// Hence numberOfSegmentsToCreate = 2^(ceil(log2(chordAngleBetweenPoints/segmentMaxLength))).
```
Comment (andyyang890) [inline pkg/geo/geogfn/segmentize.go:83]: I edited the comment for clarity and to correct the last two steps of the derivation

File: pkg/geo/geogfn/segmentize.go [hunk 4]
```diff
@@ -88,11 +93,19 @@ func segmentizeCoords(a geom.Coord, b geom.Coord, segmentMaxAngle float64) ([]fl
...
 	}
 	allSegmentizedCoordinates := make([]float64, 0, numPoints)
-	allSegmentizedCoordinates = append(allSegmentizedCoordinates, a.X(), a.Y())
+	allSegmentizedCoordinates = append(allSegmentizedCoordinates, a.Clone()...)
+	segmentFraction := 1.0 / float64(numberOfSegmentsToCreate)
 	for pointInserted := 1; pointInserted < numberOfSegmentsToCreate; pointInserted++ {
 		newPoint := s2.Interpolate(float64(pointInserted)/float64(numberOfSegmentsToCreate), pointA, pointB)
 		latLng := s2.LatLngFromPoint(newPoint)
 		allSegmentizedCoordinates = append(allSegmentizedCoordinates, latLng.Lng.Degrees(), latLng.Lat.Degrees())
+		// Linearly interpolate Z and/or M coordinates.
```
Comment (andyyang890) [inline pkg/geo/geogfn/segmentize.go:102]: Fixed the geography variant of `segmentizeCoords` as well

**Review by otan — APPROVED** [review]

**Comment by andyyang890** [comment]
bors r=otan

**Pull request merged**